torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -527,7 +527,13 @@ bool ProcessGroupAgent::handleRecv(RecvWork& work) {
       // response as a callback which fires when the future completes.
       auto fromId = work.from_.id_;
       auto requestId = work.id_;
-      futureResponse->addCallback([this, fromId, requestId, futureResponse]() {
+      futureResponse->addCallback([this,
+                                   fromId,
+                                   requestId,
+                                   weak = std::weak_ptr<FutureMessage>(
+                                       futureResponse)]() {
+        auto futureResponse = weak.lock();
+        TORCH_INTERNAL_ASSERT(futureResponse);
```

Comment:
Looks like this pattern is repeated a lot. Would it make sense to add an API like `addCallbackCaptureFuture`, where we define a Callback struct consisting of std::function (that takes Future as argument) and a weak ptr to the Future. When the callback is called we can perform weak lock check and pass the std::shared_ptr to the actual callback.

Reply:
I was thinking that, but was going back-and-forth about whether we wanted that logic in the future class. I might go ahead and do that, might be cleaner